net: ip_k66f: Rework of network stack to support raw sockets when CONFIG_NET_{UDP|TCP} enabled

lmajewski opened the pull request:
After those commits it will be possible to handle raw sockets in user programs when also UDP and TCP support is enabled.
It first moves eligible code to `conn_raw_socket()` new function and afterwards reuses it also when UDP and TCP support is
enabled.

Commit:
net: Exclude code responsible for handling raw sockets processing

The new function - namely conn_raw_socket(); has been introduced to
handle raw sockets processing. Its code, up till now, only was
executed when IS_ENABLED(CONFIG_NET_SOCKETS_PACKET) was defined.

After this change it can be reused when one would like to handle
raw sockets also when CONFIG_NET_{UDP|TCP} are enabled.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
net: Add support for simultaneous UDP/TCP and raw sockets

This patch brings support for AF_PACKET and SOCK_RAW type of sockets.
In net_conn_input() function the new flag has been introduced -
'raw_pkt_continue' to indicate if there are other than AF_PACKET
connections registered.

If we do not have other connections than AF_PACKET, the packet is
solely handled in net_conn_input() (or to be more specific in its
helper function - conn_raw_socket()).

Otherwise, it is passed back to net_conn_input in IPv4/6 processing.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

jukkar commented on subsys/net/ip/connection.c:
```diff
@@ -505,6 +505,51 @@ static bool conn_are_end_points_valid(struct net_pkt *pkt,
+		}
+
+		if (conn->cb(conn, raw_pkt, NULL,
+			     NULL, conn->user_data) ==
```
`NET_DROP) {` could be placed to this line, it does not look too long

lmajewski replied:
Fixed. I've pushed new version of those patches.

jukkar reviewed (commented):
Looks okeyish, we are missing tests to check the functionality. Could you add test(s) to `tests/net/socket/af_packet` to check couple of cases:
1. we have a packet socket and UDP socket bound, and we receive a UDP packet to bound port (in this case the data should be delivered to both sockets)
2. two sockets bound, but UDP is not listening the received data port, in this case only packet socket should receive data
at least these cases should be checked

Commit:
test: af_packet: Do not change proto to network order when creating socket

The setup_socket() function calls socket() with proto changed to network
order. In this case functions with zsock_* prefix are called instead of
zpacket_*. The problem is with 'packet_is_supported()' method from
sockets_packet.c, which returns false when ETH_P_ALL proto is converted
with htons().

This patch fixes this issue by removing the htons() call.

Signed-off-by: Lukasz Majewski <lukma@denx.de>
Suggested-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

Commit:
test: af_packet: Add test for passing UDP packet to open RAW socket

The af_packet test has been augmented to check if sent UDP packet from
one port to another (via net interface) is also passed to open SOCK_RAW
connection.
The test_packet_sockets() function has been reused to setup the SOCK_RAW
sockets for this test.
It is important to note that the packet is passed to receive part of net
stack after being sent.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

rlubos reviewed (approved):
Looks fine to me.

jukkar commented on subsys/net/ip/packet_socket.c:
```diff
@@ -35,8 +35,22 @@ enum net_verdict net_packet_socket_input(struct net_pkt *pkt)
+	enum net_verdict ret;
 
-	net_pkt_set_family(pkt, AF_PACKET);
+	if (net_pkt_family(pkt) == AF_PACKET) {
```
I just wonder where is the family set to be `AF_PACKET`. This function is called at the very beginning of the input processing, at this point we do not know the address family yet.

lmajewski replied:
I wanted to optimize the code to avoid some unnecessary cloning of packets. However, as AF_PACKET cannot be set so early the packet is cloned always anyway.  So it seems like I shall just remove the if from line 40 above.

jukkar replied:
I think we could clone the packet only if `net_conn_input()` finds out that the packet needs to be fed to UDP/TCP socket. No need to do it always if there is no socket listening those packets.

jukkar replied:
So perhaps we could feed the packet to `net_conn_input()` and if it notices that there is other than `AF_PACKET` listening data, it would return the original net_pkt (so that `NET_CONTINUE` is returned by `net_packet_socket_input()`), and let the `net_conn_input()` clone the packet and pass the cloned one to raw socket.

lmajewski replied:
I've updated patches according to your above guidelines.

jukkar commented on subsys/net/ip/connection.c:
```diff
@@ -554,6 +554,7 @@ enum net_verdict net_conn_input(struct net_pkt *pkt,
 	bool is_mcast_pkt = false, mcast_pkt_delivered = false;
```
Also the commit subject `net: Add support for SOCK_RAW and AF_PACKET` is misleading as we already have support for AF_PACKET sockets. The subject should say something like `net: Add support for simultaneous UDP and raw sockets`, also commit body could be clarified a bit.

lmajewski replied:
Ok.

lmajewski commented on subsys/net/ip/connection.c:
```diff
@@ -598,9 +639,38 @@ enum net_verdict net_conn_input(struct net_pkt *pkt,
 
 		if (conn->family != AF_UNSPEC &&
 		    conn->family != net_pkt_family(pkt)) {
+			/* When there are other connections than AF_PACKET,
+			 * the packet shall be also passed to upper layers
+			 * for handling.
```
ok

jukkar reviewed (approved):
Thanks Lukasz!
